src/compiler/core.ts

```diff
@@ -804,17 +804,16 @@ namespace ts {
     function Signature(checker: TypeChecker) {
     }
 
+    function Node(pos: number, end: number, kind: SyntaxKind) {
+        this.kind = kind;
+        this.pos = pos;
+        this.end = end;
+        this.flags = NodeFlags.None;
+        this.parent = undefined;
+    }
+
     export let objectAllocator: ObjectAllocator = {
-        getNodeConstructor: kind => {
-            function Node(pos: number, end: number) {
-                this.pos = pos;
-                this.end = end;
-                this.flags = NodeFlags.None;
-                this.parent = undefined;
-            }
-            Node.prototype = { kind };
-            return <any>Node;
-        },
+        getNodeConstructor: _ => <any>Node,

```

Comment:
It was kept as-is because on in services.ts we use different class as prototype for `SourceFile` - https://github.com/Microsoft/TypeScript/blob/master/src/services/services.ts#L7980. For this I think we can split `getNodeConstructor` into `getNodeConstructor` for regular nodes and `getSourceFileConstructor` for source files and in batch case they both will return the same constructor function.